### ISSUE TITLE
Remove resource hooks in favor of resource middleware

### ISSUE THREAD

**Issue by smyrman** [rs/rest-layer#259]
Hooks have several problems and limitations compared to middleware:
- We need two methods (before and after hooks)
- We can't control the flow as detailed as we would want (Find hooks can modify the query, Get hooks can't)
- We can't abort/change the flow without raising an error (E.g. call patch instead of delete/clear to implement soft delete).

With middleware we get more power:
- We can do both pre and post operations within a single method.
- We can control the flow, e.g. call patch instead of delete to implement soft delete.
- We can _always_ modify the query.

I want to suggest completely removing resource hooks in favor of middleware. This may involve making `resource.Resource` an interface that can be wrapped, or we might decide it's sufficient to wrap the _storage layer_ and document that as the way to implement hooks.

**Comment by rs** [comment]
The issue I see with middleware is that it makes it harder to use optional interfaces. We use those on storer for instance, IIRC.

**Comment by smyrman** [comment]
I think that's solvable with clousure defaults, however I think we can probably just drop the optional interfaces all-together and make things easier. To make writing middleware easier we can also drop the `Get` and `Delete` methods:

```go
type Storer interface{
	Count(ctx context.Context, q *query.Query) (int, error) // merged from Counter interface
	Find(ctx context.Context, q *query.Query) (*ItemList, error)
	Clear(ctx context.Context, q *query.Query) (int, error)
	Insert(ctx context.Context, items []*Item) error
	Replace(ctx context.Context, item *Item, original *query.Query) error // original as query
}
```

Middleware we could now wrote:
- Simple and safe permission checks that _only_ alter the query (for all methods).
- Soft delete can alter the flow without raising errors, e.g. call Patch on parent.

The resource could still provide convenience methods for `Delete`, `Get` and `MultiGet` by ID on top of `Clear` and `Find`. 

I am imagining we still call  a method on `Resource` to apply middleware. E.g.:

```go
// Type aliases for functions.
type (
	CountFunc = func(ctx context.Context, q *query.Query) (int, error)
	ClearFunc = CountFunc
	FindFunc = func(ctx context.Context, q *query.Query) (*ItemList, error)
	InsertFunc = func(ctx context.Context, items []*Item) error
)

type CountMiddleware struct {
	CountFunc(parent Resource) CounterFunc
}

type FindMiddleware struct {
	FindFunc(parent Resource) FindFunc
}

// ...

type Resource struct {
	count CountFunc
 	find FindFunc
	clear ClearFunc
	insert InsertFunc
	// ...
}


// newResource creates a new resource with provided spec, handler and config.
func newResource(name string, s schema.Schema, h Storer, c Conf) Resource {
	return Resource{
		count: h.Count,
		find: h.Find,
		// ...
	}
}

func (r Resource) WithMiddleware(m interface{}) Resource {
	if t, ok := m.(CountMiddleware); ok {
		r.count = t.CountFunc(r)
	}
	if t, ok := m.(FindMiddleware); ok {
		r.find = t.FindFunc(r)
	}
	// ...
	return r
}

func (r Resource) Get(id interface{}) (*Item, error) {
	// Call Find with limit 1, report 0 items found as error
}
func (r Resource) Delete(id interface{}) error {
	// Call Clear with limit 1, report 0 deleted as error
}

**Comment by rs** [comment]
Dropping `MultiGet` on storer would prevent some storage from implementing some interesting optimizations.

If you can't add a middleware for the `Get` method, how do you handle ACL for resource per id?

In your example you can not chain middleware right? Calling `WithMiddleware` more than once would override existing middleware. Also, I'm not a big fan of `m interface{}` argument.

**Comment by smyrman** [comment]
> In your example you can not chain middleware right? Calling WithMiddleware more than once would override existing middleware. Also, I'm not a big fan of m interface{} argument.

No, it can be chained; `r Resource` is not a pointer reciever, and is thus copied each time it's passed to a new variable or function.

**Comment by smyrman** [comment]
> If you can't add a middleware for the Get method, how do you handle ACL for resource per id?

> m interface{} argument

This is the same as what we get for hooks today:

```Go
// Use attaches an event handler to the resource. This event handler must
// implement on of the resource.*EventHandler interface or this method returns
// an error.
func (r *Resource) Use(e interface{}) error {
	return r.hooks.use(e)
}
```

But agree it would be nice with a concrete interface, but then you can't allow middleware for multiple operations to be passed in the same function call.

**Comment by smyrman** [comment]
> m interface{} argument

Maybe this is better:

```go
type MiddlewareFunc func(r Resource) Resource
```

We don't even need to define this, it could just be by convention.

Example middleware:
```
func WithSoftDelete(r Resource) Resource {
	parent := r
	r.Clear = func(ctx context.Context, q *query.Query) (int, error) {
		deleteTime := time.Now()
		for _, item := parent.Find(ctx, q) {
			origQ := query.Predicate{
				query.Equal{"_etag", item.Etag},
				query.Equal{"id": item.ID},
			}
			item.Payload["deletedAt"] = deleteTime
			item.UpdateEtag() // recalculate E-tag based on item changes.
			orig.Replace(item, origQ)
		}
	}
	return r
}

```
Ideally Replace would also be defined as a bulk operation somehow to allow doing this kind of things as a transaction internally in the storage layer.

**Comment by smyrman** [comment]
> Dropping MultiGet on storer would prevent some storage from implementing some interesting optimizations.

If we could get a simpler / safer access control, and get to implement soft-delete, I would say it's well worth the price.

Storage layers could _still_ optimize fetching multiple items by ID, but would need to inspect the Find predicate to do so rather than implementing an additional interfaces. I don't know which storage engine you had in mind that can do a multi-get faster than a normal find with `id :{$in: [...]}`?

**Comment by rs** [comment]
That's my point, it is expensive to introspect a query to see if an optimization is possible. Same for ACL, how do you put an ACL on `/obj/<id>` without a hook on "get"?

**Comment by smyrman** [comment]
>  Same for ACL, how do you put an ACL on /obj/<id> without a hook on "get"?

`Get` wraps Find (which already have the right middleware set).

1. Calling` Get(ctx, id)` results in a `Find(ctx, q.Query{Predicate: query.Equal{"id", id}, Window:query.Window{Limit: 1})` 
2. Find issues the relevant permission middleware to AND in the relevant permission criteria.
3. If Find returns 0 results, Get returns "not fond".

The same is through with `Delete` which would wrap a `Clear` call with all relevant middleware applied.

Our own permission syte, isn't actually ACL; permissions are stored on "protected fields" on the object, or on related objects that are fetched before the resource itself, but I don't get how this would be any different with an ACL permission system.

> That's my point, it is expensive to introspect a query to see if an optimization is possible.

For the case in particular, it's something like this:

```go
// AsIDQuery returns an IDQuery or nil.
AsIDQuery(q *query.Query) *IDQuery {
    if len(q.Predicate) != 1 {
        return nil
    }
    in, ok := q.Predicate[0].(query.In)
    if !ok || in.Field != "id" {
        return nil
    }
    return *IDQuery{IDs: in.Values, Window: q.Window}
}
```

Or something like that...

I could agree it being complex, but I don't buy that it's going to be expensive. The best thing about it complexity wise, is that the complexity is _moved_ from the rest-layer end-user to the rest-layer storer backend implementation, and that's a good thing.

To reiterate on the last point from our own experience on writing hooks, it's really hard with the current hooks. For the permission hooks in particular, we need to:
1. Modify the query on Find  / Clear / Count operations.
2. On Get, we can't modify the query (we can redirect to a Find call), and we need to wait until we have fetched the resource, and then match our permission query on the fetched item in code.

Writing a proper Soft-delete, as explained by #225, just isn't possible through hooks. Probably we _could_ do everything we want as middleware directly on top of the Storage layer, but we have so far chosen to use the documented hooks instead.

**Comment by smyrman** [comment]
> Same for ACL, how do you put an ACL on /obj/ without a hook on "get"?

OK I get what you mean now. You refer to if, e.g. a user has `get:resource` permissions but not `list:resource` permissions. I was thinking about filtering out _specific_ items that the user has permission to, and not wether or not the operation as a whole is allowed for a particular user.

Yes, you have a point. Probably we need to allow middleware for Get. As long as we can still redirect it to do a _Find_ in our per-item permission middleware, that's fine.

**Comment by rs** [comment]
Ok, if we go that route, why not simplifying it even more and just let people wrap storer to build middleware like it is down with HTTP? This way there is no complexity to be added at the rest-layer layer.

**Comment by smyrman** [comment]
Yeah, Probably thats fair.
